basis/basis.py

```diff
@@ -0,0 +1,172 @@
+import abc
+import numpy
+import math
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``max_exactness`` represents the maximum level of exactness, or 
+    accuracy, of a surrogate function this class can provide enough 
+    information for. This class can provide enough information for 
+    constructing a surrogate function of any accuracy at or below 
+    max_exactness.
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid. The Smolyak index of a given point is equal to 
+    its position in the list points.
+    ``levels`` property gives the Smolyak indexes for each
+    grid level.
+
+    Parameters
+    ----------
+    max_exactness : int
+        level of exactness class will describe
+
+    """
+
+    def __init__(self,max_exactness):
+        self._max_exactness = max_exactness
+        self._points = []
+        self._levels = []
+        self._update()
+
+    @property
+    def max_exactness(self):
+        """Maximum level of exactness described"""
+        return self._max_exactness
+
+    @property
+    def points(self):
+        """Extrema assigned to Smolyak indices"""
+        return self._points
+
+    @property
+    def levels(self):
+        """index of each point that belongs to each grid level"""
+        return self._levels
+
+    @max_exactness.setter
+    def max_exactness(self, max_exactness):
+        """Set maximum level of exactness
+        Sets maximmum level of exactness class instance will describe, will
+        extend or truncate properties points and levels to match new value
+
+        Parameters
+        ----------
+        max_exactness : int
+            new level of exactness class should describe
+        """
+        if max_exactness > self._max_exactness:
+            # extend
+            self._max_exactness = max_exactness
+            # update other properties
+            self._update()
+        elif max_exactness < self._max_exactness:
+            # truncate
+            self._max_exactness = max_exactness
+            # update other properties
+            points_keep = sum(
+                    [len(x) for x in self._levels[:self._max_exactness+1]])
+            self._points = self._points[:points_keep]
+            self._levels = self._levels[:self._max_exactness+1]
+
+
+    @abc.abstractmethod
+    def _update(self):
+        """Update properties described by level of exactness
+        Compute points of basis function and the Smolyak index of
+        each grid level between 0 and max_exactness + 1 
+        """
+        pass
```

Comment:
Yes, suggest to remove: not all basis functions need to have this method. Instead, you should leave it up to the deriving classes to decide if they need to call `_update` before they evaluate basis.
```suggestion
```